src/registry-buddy/dockerhub/client.go

```diff
@@ -0,0 +1,123 @@
+package dockerhub
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+type client struct {
+	Domain string
+}
+
+type NotFoundError struct {
+	Err error
+}
+
+func (e NotFoundError) Error() string {
+	return e.Err.Error()
+}
+
+const DefaultDomain = "https://hub.docker.com"
+
+func NewClient(domain string) *client {
+	if domain == "" {
+		domain = DefaultDomain
+	}
+	return &client{
+		Domain: domain,
+	}
+}
+
+type authorizationRequestBody struct {
+	Username string `json:"username"`
+	Password string	`json:"password"`
+}
+
+type authorizationResponseBody struct {
+	Token string `json:"token"`
+}
+
+func (c *client) GetAuthorizationToken(username, password string) (string, error) {
+	httpClient := c.buildHttpClient()
+
+	body, err := json.Marshal(authorizationRequestBody{
+		Username: username,
+		Password: password,
+	})
+	if err != nil {
+		return "", fmt.Errorf("error marshalling authorization request: %w", err) // untested
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/v2/users/login/", c.Domain), bytes.NewBuffer(body))
+	if err != nil {
+		return "", fmt.Errorf("error building authorization request: %w", err) // untested
+	}
+	req.Header.Add("Content-Type", "application/json")
+
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("error during authorization request: %w", err) // untested
+	}
+	defer res.Body.Close()
+
+	bodyBytes, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return "", fmt.Errorf("error reading response body: %w", err) // untested
+	}
+
+	if res.StatusCode == http.StatusUnauthorized {
+		return "", fmt.Errorf("unauthorized request (status %d): %s", res.StatusCode, string(bodyBytes))
+	} else if res.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("unexpected response (status %d): %s", res.StatusCode, string(bodyBytes))
+	}
+
+	resBody := authorizationResponseBody{}
+	err = json.Unmarshal(bodyBytes, &resBody)
+	if err != nil {
+		return "", fmt.Errorf("error decoding authorization response: error %w, body %q", err, string(bodyBytes)) // untested
+	}
+
+	return resBody.Token, nil
+}
+
+func (c *client) DeleteRepo(repoName, token string) error {
```

Comment:
So, I had an action item to write a story to consider making a single repo package repo _per app_ and packages would be digests under that repo. I had some concerns about longevity of a particular digest with this approach (like how many will the remote repository keep around? does it keep them indefinitely?), but those were all hypotheticals. 

Given that this is how we have to clean up on something like DockerHub I'm wondering if the way we have it now is better. 🤔

Reply:
It's also possible to delete a specific tag on Dockerhub. That's actually how the original story was written, but I then discovered we could delete the entire repo and that seemed like a cleaner solution to me. If we can't do something equivalent for Harbor, then it may make sense to switch how we store our package images for all platforms so it can be better there.